### PR TITLE
UI feature: Additional filters added to problems tab

### LINKED ISSUE #2489: UI: Provide extra filters to problems tab on environments
There have been many requests to improve the current filtering of problem results inside the 'Problem' tab on an environment.  

Some additional filtering options such as by severity, source and service would be ideal so we can better display available results.

image

### PR DESCRIPTION
**Pull request by timclifford**
# Checklist

- [x] Additional select filters for problems tab so we can improve results visibility

image

# Closing issues
closes #2489

### CODE REVIEW

File: services/ui/src/components/Problems/index.js [hunk 1]
```diff
@@ -1,229 +1,368 @@
...
+        if (item[key] !== null) {
+          return servicesSelected.indexOf(item['service'].toString()) > -1;
+        };
+    });
+  }
 
-    const handleProblemFilterChange = (event) => {
-        setHasFilter(false);
+  // Filter by select fields and text input field.
+  const filterResults = (item) => {
```
Comment (bomoko) [inline services/ui/src/components/Problems/index.js]: The "filtered" logic here is a little confusing.

On line 97 you assign "filtered" true - and then you _check_ filtered on 99, 103, and 107 - but then you _also_ set filtered there. AND used filtered as a return value.

I think that the confusion mostly comes from the naming. The name of this function is "filterResults" which initially suggested to me that this should take some results, and filter them.
But after reading through the code, I realized that this isn't actually a filter, but it's a predicate function that tells us whether or not an _item should or should not be filtered_. This slightly confusing naming is repeated in the "sourceFilter", "severityFilter" and "serviceFilter" names.

These should be changed to something more accurate, I think - something that suggests that it's going to return a boolean value. (an isX() or shouldX() function name might make sense).

File: services/ui/src/components/Problems/index.js [hunk 2]
```diff
@@ -1,229 +1,368 @@
...
+      filtered = sourceFilter(item);
+    }
+
+    if (servicesSelected.length > 0 && filtered) {
+      filtered = serviceFilter(item);
+    }
+
+    // Filter text input
+    const lowercasedFilter = problemTerm.toLowerCase();
+    if (filtered && (problemTerm != null || problemTerm !== '')) {
```
Comment (bomoko) [inline services/ui/src/components/Problems/index.js]: if you moved this into its own function (like the other three) then you could, I think, reduce this entire function to something like ...

shouldItemBeFiltered = (i) => isFilteredBySeverity(i) || isFilteredByService(i) || isFilteredbySource(i) || isFilteredByTextFilter(i)

I'm not suggesting that you do this kind of thing .. but I think we could get this a lot simpler

**Review by bomoko — CHANGES_REQUESTED** [review]

File: services/ui/src/components/Filters/helpers.js [hunk 3]
```diff
@@ -19,6 +19,12 @@ export const getSourceOptions = gql`
   }
 `;
 
+export const getServiceOptions = gql`
+  query getProblemServices($environment: Int) {
+    services: problemServices(environment: $environment)
```
[the remark this answers is not in the document]
Comment (timclifford) [reply services/ui/src/components/Filters/helpers.js]: Yep it can

File: services/ui/src/components/Problems/index.js (hunk 1 above)
Comment (timclifford) [reply services/ui/src/components/Problems/index.js]: Tidied up the naming of these now

File: services/ui/src/components/Problems/index.js (hunk 2 above)
Comment (timclifford) [reply services/ui/src/components/Problems/index.js]: moved into its own function and clean the whole function up

File: services/ui/src/components/Problems/index.js [hunk 4]
```diff
@@ -1,229 +1,368 @@
...
-        return requestSort(key);
+  const isFilteredByTextFilter = (item) => {
+    return Object.keys(item).some(key => {
+      if (item[key] !== null) {
+        return item[key].toString().toLowerCase().includes(problemTerm.toLowerCase());
+      }
+    });
+  }
+
+  const shouldItemBeShown = (item) => {
```
Comment (bomoko) [inline services/ui/src/components/Problems/index.js:112]: Hi @timclifford 

As we discussed earlier - if you move the various checks you're doing here in the "if" statements' conditionals, we can simplify the logic of this function to something like (with the functions renamed)

const shouldItemBeShown = (item) => matchesSeveritySelector(item) && matchesServiceSelector(item) && matchesSourceSelector(item) && matchesTextFilter(item)

**Review by bomoko — CHANGES_REQUESTED** [review]

File: services/ui/src/components/Problems/index.js (hunk 4 above)
Comment (timclifford) [reply services/ui/src/components/Problems/index.js:112]: Thanks again for looking at this - as discussed, renamed functions to something a little clearer and improved the conditional logic too!

**Review by bomoko — APPROVED** [review]

**Review by bomoko — APPROVED** [review]
Thanks for the changes, @timclifford - lgtm

**Comment by tobybellwood** [comment]
@bomoko and @timclifford - if this needs to make it to Lagoon in the next few weeks, this PR will need to be rebased on the `master` branch instead of `main` before we release 1.14 (as 1.14 is a Lagoon 1-only release)

**Comment by timclifford** [comment]
> @bomoko and @timclifford - if this needs to make it to Lagoon in the next few weeks, this PR will need to be rebased on the `master` branch instead of `main` before we release 1.14 (as 1.14 is a Lagoon 1-only release)

Hi @tobybellwood As discussed, I re-branched off of master with the same changes as above into a new PR #2526

**Pull request merged**